[SDK Docs Audit] Swift Geofences feature improvements

hokstuff opened the pull request:
<!--
Thank you for contributing to Braze Docs! Read comments like this one for help filling out your Pull Request (PR). For more information about contributing to Braze Docs, see:
braze.com/docs/contributing/.
-->

### Why are you making this change? (required)
<!-- In one or two sentences, describe your changes. -->
Internal team observations from documentation audit initiative - Swift Geofences

### Related PRs, issues, or features (optional)
<!--
For related GitHub issues and features, add "Fixes #ISSUE_NUMBER".
For Jira tickets, add the ticket number and a link to the ticket, like:
[BD-NUMBER](https://jira.braze.com/browse/BD-NUMBER)
-->
- https://jira.braze.com/browse/SDK-5680

### Feature release date (optional)
<!-- If this PR is related to an upcoming Braze feature, add the date below, like: 15 May 2024. -->
- N/A

### Contributor checklist
<!-- After you've filled out the previous sections, select "Create Draft Pull Request", then review your PR before filling out the following checklist. -->
- [x] I confirm that my PR meets the following:
    - I signed the [Contribution License Agreement (CLA)](https://www.braze.com/docs/cla).
    - My style and voice follows the [Braze Style Guide](https://docs.google.com/document/u/2/d/e/2PACX-1vTluyDFO3ZEV7V6VvhXE4As_hSFwmnFFdU9g6_TrAYTgH1QmbRoEDDdn5GzKAB9vdBbIdyiFdoaJcNk/pub).
    - My content contains correct spelling and grammar.
    - All links are working correctly.
    - If I renamed or moved a file or directory, I set up [URL redirects](https://www.braze.com/docs/contributing/content_management/redirecting_urls) for each file.
    - If I updated or replaced an image, I did not remove the original image file from the repository. (For more information, see [Updating an image](https://www.braze.com/docs/contributing/content_management/images/#updating-an-image)).
    - If my PR is related to a paid SKU, third party, SMS, AI, or privacy, I have received written approval from Braze Legal.

### Submitting for review

If your PR meets the above requirements, select **Ready for review**, then add a reviewer:
- **Non-Braze contributors:** Tag `@braze-inc/docs-team` in a comment below.
- **Braze employees:** Have any relevant subject matter experts (like engineers or product managers) review your work first, then add the [tech writer assigned to your specific vertical](https://confluence.braze.com/pages/viewpage.action?pageId=75039896). If you're not sure which tech writer to add, you can add `braze-inc/docs-team` instead. 

Thanks for contributing! We look forward to reading your work.

hokstuff commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -58,16 +54,62 @@ AppDelegate.braze = braze;
+
+However, you can choose to monitor geofence events when your app is in the background or if [`When In Use` authorization](#swift_request-authorization) is active. To do this, add the `Background Mode -> Location updates` capability to your Xcode project.
+
+![In Xcode, Background Mode > Location Updates]({% image_buster /assets/img/push_implementation_guide/xcode-background-modes-location-updates.png %})
```
```suggestion
![In Xcode, Background Mode > Location Updates]({% image_buster /assets/img_archive/xcode-background-modes-location-updates.png %})
```

jacksonemiller commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -4,15 +4,15 @@ As of iOS 14, geofences do not work reliably for users who choose to only give t
 
 ### Step 1: Enable location services
 
 Braze location services [must be enabled](https://braze-inc.github.io/braze-swift-sdk/tutorials/braze/d1-brazelocation/) through the SDK. They are not enabled by default.
```
In my opinion/experience, the steps listed there are pretty much the same steps listed below except that they also link to an Apple-hosted guide on setting up location services. I kind of think the redundancy could be somewhat confusing. I also think the aforementioned Apple guide actually isn't that great, if we have any interest in trying to host that info here.

hokstuff replied:
Yeah it has been a concern that we may have docs in 2 separate places (public docs, and our Swift tutorials). Are you suggesting that we may want to add the instructions directly into the public docs here, rather than have the integrator bounce back and forth b/t our two docs places?

jacksonemiller replied:
I more so think that what's being linked to here should be presented as an alternative, more code-centric tutorial rather then as pre-requisite steps for the rest of this tutorial.

jacksonemiller commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -4,15 +4,15 @@ As of iOS 14, geofences do not work reliably for users who choose to only give t
 
 {% multi_lang_include developer_guide/prerequisites/swift.md %} Additionally, you'll need to [set up silent push notifications]({{site.baseurl}}/developer_guide/push_notifications/silent/?sdktab=swift).
```
I was able to get away without reading the silent push notification article? I did thought read the regular push set-up article. My thought would be that the silent push tutorial is more involved than is necessary for enabling geofences.

hokstuff replied:
I believe parts of the Geofence feature can work without silent push integrated. However, I think Geofence syncs are sent up via silent push, so that might be why we say it is a prerequisite for Geofences

[These methods](https://github.com/braze-inc/swift-sdk/blob/develop/Sources/BrazeKit/_BrazeCore/Core/Notification/Payload/NotificationPayload.swift#L150-L159) are indirectly mentioned [in the silent push public docs](https://www.braze.com/docs/developer_guide/push_notifications/silent#silent-push-notifications:~:text=.Notifications.-,isInternalNotification,-(userInfo)) as well

jacksonemiller commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -218,3 +262,19 @@ AppDelegate.braze?.requestGeofences(latitude: latitude, longitude: longitude)
 {% endtab %}
 {% endtabs %}
+
+## Frequently Asked Questions (FAQ) {#faq}
```
I think we should add a new section about interactions between Braze's geofence feature and any other geofence usage the customer might have going on. I notice many PQs about this. I can push something later this week.

hokstuff replied:
What types of interactions are you thinking about adding? Also feel free to just push what you were thinking

jacksonemiller commented:
I left some feedback after successfully integrating geofences into a sample app. I can help implement some of it later this week. I will say that for me personally, the biggest stumper was realizing that I needed to enable geofences in the app settings page for my sample app 😅. Maybe we don't expect that job to fall to the developers 100% of the time?

hokstuff commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -218,3 +262,19 @@ AppDelegate.braze?.requestGeofences(latitude: latitude, longitude: longitude)
 
 {% endtab %}
 {% endtabs %}
+
```
Reminder: Add an FAQ to recommend integrators to use the Braze Debugger to help debug this feature if they entered/exited a Geofence

cc @torossian97

internetisaiah replied:
we currently have this noted throughout a few dev guide articles:

```markdown
{% alert note %}
While performing QA on your SDK integration, use the [SDK Debugger]({{site.baseurl}}/developer_guide/sdk_integration/debugging) to get troubleshoot issues without turning on verbose logging for your app.
{% endalert %}
```

but i'm thinking we can move it to an include + clean up the language (if needed) and then link it in this articles where applicable?

cc @torossian97 @hokstuff

hokstuff replied:
This was a specific Product team ask to include a link to use the SDK debugger especially for Geofences due to the high Support volume. But yes I'm down to clean up the language and have something link out from this section!

hokstuff commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -4,15 +4,15 @@ As of iOS 14, geofences do not work reliably for users who choose to only give t
 
-## Setting up geofences
+## Setting up geofences {#setting-up-geofences}
 
```
Note: We should add instructions (+ a screenshot because clients can't miss those 😂) to enable the Geofences toggle in the Dashboard, [like shown here](https://brazetechnology.slack.com/archives/G0171TAJJ1L/p1746546448992179)

I think to make sure it's not forgotten, we can also add this to the complementary Android docs while we're at it

jerielng commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -218,3 +262,19 @@ AppDelegate.braze?.requestGeofences(latitude: latitude, longitude: longitude)
+
+#### Why are my geofences with Braze not being received by my device?
+
+The iOS operating system only allows a single app to store a maximum of 20 geofences. If your app also stores non-Braze geofences, there may not be space for the Braze geofences.
```
For my knowledge, what does storage mean in this case? Is there a clearance policy, or rather, how do the 20 geofences get refreshed when appropriate?

jerielng replied:
That sounds good to me - I think we can probably keep it as it is now!

jerielng commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -218,3 +262,19 @@ AppDelegate.braze?.requestGeofences(latitude: latitude, longitude: longitude)
+
+#### Can the Geofences feature be used while a device is offline?
+
+If the device was connected to the internet and successfully receives geofences from the server, it is possible to enter a geofence even if the device is offline. This is because a device's location operates separately from its internet connectivity.
```
Editing for flow. Let me know what you think of this:
```suggestion
A device needs to be connected to the internet only when a refresh occurs. Once it has successfully received geofences from the server, it is possible to enter a geofence even if the device is offline. This is because a device's location operates separately from its internet connectivity.
```

jerielng commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -110,7 +152,9 @@ CLLocationManager *locationManager = [[CLLocationManager alloc] init];
-By default, `requestAlwaysAuthorization()` only grants your app `When In Use` authorization and will re-prompt your user for `Always` authorization after some time has passed. However, you can choose to immediately prompt your user by first calling `requestWhenInUseAuthorization()`, then calling `requestAlwaysAuthorization()` after receiving your initial `When In Use` authorization.
+By default, `requestAlwaysAuthorization()` only grants your app `When In Use` authorization and will re-prompt your user for `Always` authorization after some time has passed.
+
+However, you can choose to immediately prompt your user by first calling `requestWhenInUseAuthorization()`, then calling `requestAlwaysAuthorization()` after receiving your initial `When In Use` authorization.
```
```suggestion
However, you can choose to immediately prompt your user by first calling `requestWhenInUseAuthorization()` and then calling `requestAlwaysAuthorization()` after receiving your initial `When In Use` authorization.
```

jerielng commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -195,11 +239,11 @@ AppDelegate.braze = braze;
+### Step 2: Call `requestGeofences` manually
 
-When the Braze SDK requests geofences to monitor from the backend, it reports the user's current location and receives geofences that are determined to be optimally relevant based on the location reported. There is a rate limit of one geofence refresh per session.
+When the Braze SDK requests geofences from the backend, it reports the user's current location and receives geofences that are determined to be optimally relevant based on the location reported.
```
Should we move these two paragraphs to the top-level section, right before step 1? This information seems relevant for the entire section as a whole. When I got to this section, it wasn't clear to me why we would want to do this until I read step 2.

jerielng reviewed (commented):
Open question: Should we document what circumstances will trigger the SDK to automatically request geofences?

hokstuff commented:
> Open question: Should we document what circumstances will trigger the SDK to automatically request geofences?

I think it's reasonable (nice to have?) to add this to the FAQ if it is deterministic, though I think it would be beneficial to also include that info under Android to avoid Support Qs

hokstuff commented:
I made quite a handful of changes, including:
1. Moved the step to configure the Braze Dashboard as Step 1, instead of Step 6
2. Moved more content into the FAQ section
3. Extracted Swift tutorial steps into this page directly
4. Address all remaining open comments above

I'm going to re-request reviews now

jacksonemiller commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -218,3 +274,33 @@ AppDelegate.braze?.requestGeofences(latitude: latitude, longitude: longitude)
+
+As mentioned above, iOS allows a single app to store a maximum of 20 geofences. This storage is shared by both Braze and non-Braze geofences and is managed by [CLLocationManager](https://developer.apple.com/documentation/corelocation/cllocationmanager).
+
+For instance, if your app contains 20 non-Braze geofences, there would be no storage to track any Braze geofences. In order to receive new Braze geofences, you will need to use [Apple's location APIs](https://developer.apple.com/documentation/corelocation) to stop monitoring some of the existing geofences on the device.
```
Double check me on this, but isn't it equally possible that the Braze geofences kick out some of the customer's own geofences?

hokstuff replied:
Yes it is equally possible for that too. I put "for example" just to highlight this case. I guess I can add "or vise versa" if we want to be more explicit in case the reader doesn't understand it can happen the opposite way

jerielng commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -2,17 +2,60 @@
+
+#### Add the `BrazeLocation` module
+
+In the `General` tab of your `Application` configuration page, under `Frameworks and Libraries`, add the `BrazeLocation` module.
```
```suggestion
In the `General` tab of your `Application` configuration page, under `Frameworks, Libraries, & Embedded Content`, add the `BrazeLocation` module.
```

jerielng reviewed (approved):
Updates LGTM! Thanks for the deep dive - this was great

hokstuff commented on assets/img_archive/add-brazeLocation-module-xcode.png:
@internetisaiah I added this image and the one below to the `assets/img_archive/*` for convenience, but feel free to move it to the appropriate folder if we shouldn't be adding new images to the archive. Thanks!

internetisaiah commented on _includes/developer_guide/android/geofences.md:
```diff
@@ -2,7 +2,38 @@
-### Step 1: Update `build.gradle`
+### Step 1: Configure the Braze dashboard
+
+There are two ways to enable geofences for a particular app: from the **Locations** page or from the **App Settings** page.
```
ooo, love me some tabs--nice!

internetisaiah commented on _includes/developer_guide/swift/geofences.md:
```diff
@@ -2,17 +2,61 @@
+
+#### Step 2.2: Update your `Info.plist`
+
+In your `info.plist`, assign a `String` value to one of the following keys that describes why your application needs to track location. This string will be shown when your users are prompted for location services, so be sure to clearly explain the value of enabling this feature for your app.
```
@hokstuff , could we get an example code block here? can be very bare bones and nothing fancy. 👍🏽

hokstuff replied:
There is no sample code that we can provide, but we can add this screenshot in so they know what a sample description would look like in Xcode! Apple takes into account the description when approving apps for the App Store so it needs to be a legit description, which is why we didn't want to have a generic one like in the screenshot

<img width="672" alt="Screenshot 2025-05-08 at 4 58 29 PM" src="https://github.com/user-attachments/assets/9e693329-b194-434e-9e43-edad1e62e5ce" />

hokstuff replied:
It is the same text that appears when prompting the user:
<img src="https://github.com/user-attachments/assets/45508a48-d1ee-49fe-812f-f5a817841329" height=400>

internetisaiah replied:
ohhh, got it. its not done in the code its done in Xcode?

hokstuff replied:
Yep it's done in the Xcode UI instead of in code:
<img width="1104" alt="Screenshot 2025-05-08 at 6 02 17 PM" src="https://github.com/user-attachments/assets/6d8b22cc-311f-4f42-86aa-f70bf661717a" />

hokstuff commented on _includes/developer_guide/android/geofences.md:
```diff
@@ -2,7 +2,38 @@
+Android only allows up to 100 geofences to be stored for a given app. Braze locations products will use up to 20 geofence slots if available. To prevent accidental or unwanted disruption to other geofence-related functionality in your app, location geofences must be enabled for individual apps on the dashboard.
+
+For Braze locations products to work correctly, confirm that your app is not using all available geofence spots.
+{% endalert %}
```
Just noting so we don't forget - These Android tabs are a copy-paste of the Swift step, so if you make changes to the Swift side, please port them here too 🙏

internetisaiah replied:
fyi: moved info to an include to make it easier to maintain

internetisaiah reviewed (approved):
LGTM
